lib/libz/compress.3
.\"	$OpenBSD: compress.3,v 1.14 2013/07/16 15:21:11 schwarze Exp $
.\"
.\" Copyright (C) 1995-2005 Jean-loup Gailly and Mark Adler
.\"
.\" This software is provided 'as-is', without any express or implied
.\" warranty. In no event will the authors be held liable for any damages
.\" arising from the use of this software.
.\"
.\" Permission is granted to anyone to use this software for any purpose,
.\" including commercial applications, and to alter it and redistribute it
.\" freely, subject to the following restrictions:
.\"
.\" The origin of this software must not be misrepresented; you must not
.\" claim that you wrote the original software. If you use this software
.\" in a product, an acknowledgment in the product documentation would be
.\" appreciated but is not required.
.\" Altered source versions must be plainly marked as such, and must not be
.\" misrepresented as being the original software.
.\" This notice may not be removed or altered from any source distribution.
.\"
.\" Converted to mdoc format for the OpenBSD project
.\" by Jason McIntyre <jmc@openbsd.org>
.\"
.\" This page corresponds to zlib version 1.2.8
.\"
.Dd $Mdocdate: January 12 2014 $
.Dt COMPRESS 3
.Os
.Sh NAME
.Nm compress
.Nd zlib general purpose compression library
.Sh SYNOPSIS
.In zlib.h
.Pp
Basic functions
.Pp
.Ft const char *
.Fn zlibVersion "void"
.Ft int
.Fn deflateInit "z_streamp strm" "int level"
.Ft int
.Fn deflate "z_streamp strm" "int flush"
.Ft int
.Fn deflateEnd "z_streamp strm"
.Ft int
.Fn inflateInit "z_streamp strm"
.Ft int
.Fn inflate "z_streamp strm" "int flush"
.Ft int
.Fn inflateEnd "z_streamp strm"
.Pp
Advanced functions
.Pp
.Ft int
.Fn deflateInit2 "z_streamp strm" "int level" "int method" "int windowBits" "int memLevel" "int strategy"
.Ft int
.Fn deflateSetDictionary "z_streamp strm" "const Bytef *dictionary" "uInt dictLength"
.Ft int
.Fn deflateCopy "z_streamp dest" "z_streamp source"
.Ft int
.Fn deflateReset "z_streamp strm"
.Ft int
.Fn deflateParams "z_streamp strm" "int level" "int strategy"
.Ft int
.Fn deflateTune "z_streamp strm" "int good_length" "int max_lazy" "int nice_length" "int max_chain"
.Ft uLong
.Fn deflateBound "z_streamp strm" "uLong sourceLen"
.Ft int
.Fn deflatePending "z_streamp strm" "unsigned *pending" "int *bits"
.Ft int
.Fn deflatePrime "z_streamp strm" "int bits" "int value"
.Ft int
.Fn deflateSetHeader "z_streamp strm" "gz_headerp head"
.Ft int
.Fn inflateInit2 "z_streamp strm" "int windowBits"
.Ft int
.Fn inflateSetDictionary "z_streamp strm" "const Bytef *dictionary" "uInt dictLength"
.Ft int
.Fn inflateGetDictionary "z_streamp strm" "Bytef *dictionary" "uInt *dictLength"
.Ft int
.Fn inflateSync "z_streamp strm"
.Ft int
.Fn inflateCopy "z_streamp dst" "z_streamp source"
.Ft int
.Fn inflateReset "z_streamp strm"
.Ft int
.Fn inflateReset2 "z_streamp strm" "int windowBits"
.Ft int
.Fn inflatePrime "z_streamp strm" "int bits" "int value"
.Ft int
.Fn inflateMark "z_streamp strm"
.Ft int
.Fn inflateGetHeader "z_streamp strm" "gz_headerp head"
.Ft int
.Fn inflateBackInit "z_stream *strm" "int windowBits" "unsigned char FAR *window"
.Ft int
.Fn inflateBack "z_stream *strm" "in_func in" "void FAR *in_desc" "out_func out" "void FAR *out_desc"
.Ft int
.Fn inflateBackEnd "z_stream *strm"
.Ft uLong
.Fn zlibCompileFlags "void"
.Pp
Utility functions
.Pp
.Fd typedef struct gzFile_s *gzFile;
.Pp
.Ft int
.Fn compress "Bytef *dest" "uLongf *destLen" "const Bytef *source" "uLong sourceLen"
.Ft int
.Fn compress2 "Bytef *dest" "uLongf *destLen" "const Bytef *source" "uLong sourceLen" "int level"
.Ft uLong
.Fn compressBound "uLong sourceLen"
.Ft int
.Fn uncompress "Bytef *dest" "uLongf *destLen" "const Bytef *source" "uLong sourceLen"
.Pp
.Xr gzip 1
file access functions
.Pp
.Ft gzFile
.Fn gzopen "const char *path" "const char *mode"
.Ft gzFile
.Fn gzdopen "int fd" "const char *mode"
.Ft int
.Fn gzbuffer "gzFile file" "unsigned size"
.Ft int
.Fn gzsetparams "gzFile file" "int level" "int strategy"
.Ft int
.Fn gzread "gzFile file" "voidp buf" "unsigned len"
.Ft int
.Fn gzwrite "gzFile file" "voidpc buf" "unsigned len"
.Ft int
.Fn gzprintf "gzFile file" "const char *format" "..."
.Ft int
.Fn gzputs "gzFile file" "const char *s"
.Ft char *
.Fn gzgets "gzFile file" "char *buf" "int len"
.Ft int
.Fn gzputc "gzFile file" "int c"
.Ft int
.Fn gzgetc "gzFile file"
.Ft int
.Fn gzungetc "int c" "gzFile file"
.Ft int
.Fn gzflush "gzFile file" "int flush"
.Ft z_off_t
.Fn gzseek "gzFile file" "z_off_t offset" "int whence"
.Ft int
.Fn gzrewind "gzFile file"
.Ft z_off_t
.Fn gztell "gzFile file"
.Ft z_off_t
.Fn gzoffset "gzFile file"
.Ft int
.Fn gzeof "gzFile file"
.Ft int
.Fn gzdirect "gzFile file"
.Ft int
.Fn gzclose "gzFile file"
.Ft int
.Fn gzclose_r "gzFile file"
.Ft int
.Fn gzclose_w "gzFile file"
.Ft const char *
.Fn gzerror "gzFile file" "int *errnum"
.Ft void
.Fn gzclearerr "gzFile file"
.Pp
Checksum functions
.Pp
.Ft uLong
.Fn adler32 "uLong adler" "const Bytef *buf" "uInt len"
.Ft uLong
.Fn adler32_combine "uLong adler1" "uLong adler2" "z_off_t len2"
.Ft uLong
.Fn crc32 "uLong crc" "const Bytef *buf" "uInt len"
.Ft uLong
.Fn crc32_combine "uLong crc1" "uLong crc2" "z_off_t len2"
.Sh DESCRIPTION
This manual page describes the
.Nm zlib
general purpose compression library, version 1.2.8.
.Pp
The
.Nm zlib
compression library provides in-memory compression and decompression functions,
including integrity checks of the uncompressed data. This version of the
library supports only one compression method
.Pq deflation
but other algorithms will be added later and will have the same stream
interface.
.Pp
Compression can be done in a single step if the buffers are large enough, or
can be done by repeated calls of the compression function. In the latter case,
the application must provide more input and/or consume the output
.Pq providing more output space
before each call.
.Pp
The compressed data format used by default by the in-memory functions is the
.Nm zlib
format, which is a zlib wrapper documented in RFC 1950, wrapped around a
deflate stream, which is itself documented in RFC 1951.
.Pp
The library also supports reading and writing files in
.Xr gzip 1
.Pq .gz
format with an interface similar to that of
.Xr stdio 3
using the functions that start with
.Qq gz .
The gzip format is different from the zlib format. gzip is a gzip wrapper,
documented in RFC 1952, wrapped around a deflate stream. This library can
optionally read and write gzip streams in memory as well.
.Pp
The zlib format was designed to be compact and fast for use in memory and on
communications channels. The gzip format was designed for single-file
compression on file systems, has a larger header than zlib to maintain
directory information, and uses a different, slower, check method than zlib.
.Pp
The library does not install any signal handler. The decoder checks the
consistency of the compressed data, so the library should never crash even in
case of corrupted input.
.Pp
The functions within the library are divided into the following sections:
.Pp
.Bl -dash -offset indent -compact
.It
Basic functions
.It
Advanced functions
.It
Utility functions
.It
.Xr gzip 1
file access functions
.It
Checksum functions
.El
.Sh BASIC FUNCTIONS
.Bl -tag -width Ds
.It Xo
.Ft const char *
.Fn zlibVersion "void" ;
.Xc
.Pp
The application can compare
.Fn zlibVersion
and
.Dv ZLIB_VERSION
for consistency. If the first character differs, the library code actually used
is not compatible with the
.Aq Pa zlib.h
header file used by the application. This check is automatically made by
.Fn deflateInit
and
.Fn inflateInit .
.It Xo
.Ft int
.Fn deflateInit "z_streamp strm" "int level" ;
.Xc
.Pp
The
.Fn deflateInit
function initializes the internal stream state for compression. The fields
.Va zalloc ,
.Va zfree ,
and
.Va opaque
must be initialized before by the caller. If
.Va zalloc
and
.Va zfree
are set to
.Dv Z_NULL ,
.Fn deflateInit
updates them to use default allocation functions.
.Pp
The compression level must be
.Dv Z_DEFAULT_COMPRESSION ,
or between 0 and 9:
1 gives best speed, 9 gives best compression, 0 gives no compression at all
.Pq the input data is simply copied a block at a time .
.Pp
.Dv Z_DEFAULT_COMPRESSION
requests a default compromise between speed and compression
.Pq currently equivalent to level 6 .
.Pp
.Fn deflateInit
returns
.Dv Z_OK
if successful,
.Dv Z_MEM_ERROR
if there was not enough memory,
.Dv Z_STREAM_ERROR
if
.Fa level
is not a valid compression level, or
.Dv Z_VERSION_ERROR
if the
.Nm zlib
library version
.Pq zlib_version
is incompatible with the version assumed by the caller
.Pq ZLIB_VERSION .
.Va msg
is set to null if there is no error message.
.Fn deflateInit
does not perform any compression: this will be done by
.Fn deflate .
.It Xo
.Ft int
.Fn deflate "z_streamp strm" "int flush" ;
.Xc
.Pp
.Fn deflate
compresses as much data as possible, and stops when the input buffer becomes
empty or the output buffer becomes full. It may introduce some output latency
.Pq reading input without producing any output
except when forced to flush.
.Pp
The detailed semantics are as follows.
.Fn deflate
performs one or both of the following actions:
.Pp
Compress more input starting at
.Va next_in
and update
.Va next_in
and
.Va avail_in
accordingly.
If not all input can be processed
.Pq because there is not enough room in the output buffer ,
.Va next_in
and
.Va avail_in
are updated and processing will resume at this point for the next call to
.Fn deflate .
.Pp
Provide more output starting at
.Va next_out
and update
.Va next_out
and
.Va avail_out
accordingly. This action is forced if the parameter
.Fa flush
is non-zero. Forcing
.Fa flush
frequently degrades the compression ratio, so this parameter should be set only
when necessary
.Pq in interactive applications .
Some output may be provided even if
.Fa flush
is not set.
.Pp
Before the call to
.Fn deflate ,
the application should ensure that at least one of the actions is possible, by
providing more input and/or consuming more output, and updating
.Va avail_in
or
.Va avail_out
accordingly;
.Va avail_out
should never be zero before the call. The application can consume the
compressed output when it wants, for example when the output buffer is full
.Pq avail_out == 0 ,
or after each call to
.Fn deflate .
If
.Fn deflate
returns
.Dv Z_OK
and with zero
.Va avail_out ,
it must be called again after making room in the output buffer because there
might be more output pending.
.Pp
Normally the parameter
.Fa flush
is set to
.Dv Z_NO_FLUSH ,
which allows
.Fn deflate
to decide how much data to accumulate before producing output, in order to
maximise compression.
.Pp
If the parameter
.Fa flush
is set to
.Dv Z_SYNC_FLUSH ,
all pending output is flushed to the output buffer and the output is aligned on
a byte boundary, so that the decompressor can get all input data available so
far.
.Po
In particular,
.Va avail_in
is zero after the call if enough output space has been provided before the
call.
.Pc
Flushing may degrade compression for some compression algorithms and so it
should be used only when necessary. This completes the current deflate block
and follows it with an empty stored block that is three bits plus filler bits
to the next byte, followed by four bytes
.Pq 00 00 ff ff .
.Pp
If flush is set to
.Dv Z_PARTIAL_FLUSH ,
all pending output is flushed to the output buffer, but the output is not
aligned to a byte boundary. All of the input data so far will be available to
the decompressor, as for
.Dv Z_SYNC_FLUSH .
This completes the current deflate block and follows it with an empty fixed
codes block that is 10 bits long. This assures that enough bytes are output in
order for the decompressor to finish the block before the empty fixed code
block.
.Pp
If flush is set to
.Dv Z_BLOCK ,
a deflate block is completed and emitted, as for
.Dv Z_SYNC_FLUSH ,
but the output is not aligned on a byte boundary, and up to seven bits of the
current block are held to be written as the next byte after the next deflate
block is completed. In this case, the decompressor may not be provided enough
bits at this point in order to complete decompression of the data provided so
far to the compressor. It may need to wait for the next block to be emitted.
This is for advanced applications that need to control the emission of deflate
blocks.
.Pp
If
.Fa flush
is set to
.Dv Z_FULL_FLUSH ,
all output is flushed as with
.Dv Z_SYNC_FLUSH ,
and the compression state is reset so that decompression can restart from this
point if previous compressed data has been damaged or if random access is
desired. Using
.Dv Z_FULL_FLUSH
too often can seriously degrade compression.
.Pp
If
.Fn deflate
returns with avail_out == 0, this function must be called again with the same
value of the
.Fa flush
parameter and more output space
.Po
updated
.va avail_out
.Pc ,
until the flush is complete
.Po
.Fn deflate
returns with non-zero
.Va avail_out
.Pc .
In the case of a
.Dv Z_FULL_FLUSH
or a
.Dv Z_SYNC_FLUSH ,
make sure that
.Va avail_out
is greater than six to avoid repeated flush markers due to avail_out == 0 on
return.
.Pp
If the parameter
.Fa flush
is set to
.Dv Z_FINISH ,
pending input is processed, pending output is flushed and
.Fn deflate
returns with
.Dv Z_STREAM_END
if there was enough output space; if
.Fn deflate
returns with
.Dv Z_OK ,
this function must be called again with
.Dv Z_FINISH
and more output space
.Po
updated
.Va avail_out
.Pc
but no more input data, until it returns with
.Dv Z_STREAM_END
or an error.
After
.Fn deflate
has returned
.Dv Z_STREAM_END ,
the only possible operations on the stream are
.Fn deflateReset
or
.Fn deflateEnd .
.Pp
.Dv Z_FINISH
can be used immediately after
.Fn deflateInit
if all the compression is to be done in a single step.
In this case,
.Va avail_out
must be at least the value returned by
.Fn deflateBound
.Pq see below .
Then deflate is guaranteed to return Z_STREAM_END. If not enough output space
is provided,
.Fn deflate
will not return
.Dv Z_STREAM_END ,
and it must be called again as described above.
.Pp
.Fn deflate
sets strm-\*(Gtadler to the Adler-32 checksum of all input read so far
.Po
that is,
.Va total_in
bytes
.Pc .
.Pp
.Fn deflate
may update strm-\*(Gtdata_type
if it can make a good guess about the input data type
.Pq Z_BINARY or Z_TEXT .
If in doubt, the data is considered binary.
This field is only for information purposes and does not affect
the compression algorithm in any manner.
.Pp
.Fn deflate
returns
.Dv Z_OK
if some progress has been made
.Pq more input processed or more output produced ,
.Dv Z_STREAM_END
if all input has been consumed and all output has been produced
.Po
only when
.Fa flush
is set to
.Dv Z_FINISH
.Pc ,
.Dv Z_STREAM_ERROR
if the stream state was inconsistent
.Po
for example, if
.va next_in
or
.Va next_out
was
.Dv Z_NULL
.Pc ,
.Dv Z_BUF_ERROR
if no progress is possible
.Po
for example,
.Va avail_in
or
.Va avail_out
was zero
.Pc .
Note that
.Dv Z_BUF_ERROR
is not fatal, and
.Fn deflate
can be called again with more input and more output space
to continue processing.
.It Xo
.Ft int
.Fn deflateEnd "z_streamp strm" ;
.Xc
.Pp
All dynamically allocated data structures for this stream are freed.
This function discards any unprocessed input and does not flush any
pending output.
.Pp
.Fn deflateEnd
returns
.Dv Z_OK
if successful,
.Dv Z_STREAM_ERROR
if the stream state was inconsistent,
.Dv Z_DATA_ERROR
if the stream was freed prematurely
.Pq some input or output was discarded .
In the error case,
.Va msg
may be set but then points to a static string
.Pq which must not be deallocated .
.It Xo
.Ft int
.Fn inflateInit "z_streamp strm" ;
.Xc
The
.Fn inflateInit
function initializes the internal stream state for decompression.
The fields
.Va next_in ,
.Va avail_in ,
.Va zalloc ,
.Va zfree ,
and
.Va opaque
must be initialized before by the caller.
If
.Va next_in
is not
.Dv Z_NULL
and
.Va avail_in
is large enough
.Pq the exact value depends on the compression method ,
.Fn inflateInit
determines the compression method from the
.Nm zlib
header and allocates all data structures accordingly;
otherwise the allocation will be deferred to the first call to
.Fn inflate .
If
.Va zalloc
and
.Va zfree
are set to
.Dv Z_NULL ,
.Fn inflateInit
updates them to use default allocation functions.
.Pp
.Fn inflateInit
returns
.Dv Z_OK
if successful,
.Dv Z_MEM_ERROR
if there was not enough memory,
.Dv Z_VERSION_ERROR
if the
.Nm zlib
library version is incompatible with the version assumed by the caller, or
.Dv Z_STREAM_ERROR
if the parameters are invalid, such as a null pointer to the structure.
.Va msg
is set to null if there is no error message.
.Fn inflateInit
does not perform any decompression apart from possibly reading the
.Nm zlib
header if present: actual decompression will be done by
.Fn inflate .
.Po
So
.Va next_in
and
.Va avail_in
may be modified,
but
.Va next_out
and
.Va avail_out
are unused and unchanged.
.Pc
The current implementation of
.Fn inflateInit
does not process any header information -- that is deferred until
.Fn inflate
is called.
.It Xo
.Ft int
.Fn inflate "z_streamp strm" "int flush" ;
.Xc
.Fn inflate
decompresses as much data as possible, and stops when the input
buffer becomes empty or the output buffer becomes full.
It may introduce some output latency
.Pq reading input without producing any output
except when forced to flush.
.Pp
The detailed semantics are as follows.
.Fn inflate
performs one or both of the following actions:
.Pp
Decompress more input starting at
.Va next_in
and update
.Va next_in
and
.Va avail_in
accordingly.
If not all input can be processed
.Pq because there is not enough room in the output buffer ,
.Va next_in
is updated and processing will resume at this point for the next call to
.Fn inflate .
.Pp
Provide more output starting at
.Va next_out
and update
.Va next_out
and
.Va avail_out
accordingly.
.Fn inflate
provides as much output as possible,
until there is no more input data or no more space in the output buffer
.Pq see below about the flush parameter .
.Pp
Before the call to
.Fn inflate ,
the application should ensure that at least one of the actions is possible,
by providing more input and/or consuming more output,
and updating the next_* and avail_* values accordingly.
The application can consume the uncompressed output when it wants,
for example when the output buffer is full
.Pq avail_out == 0 ,
or after each call to
.Fn inflate .
If
.Fn inflate
returns
.Dv Z_OK
and with zero
.Va avail_out ,
it must be called again after making room
in the output buffer because there might be more output pending.
.Pp
The
.Fa flush
parameter of
.Fn inflate
can be
.Dv Z_NO_FLUSH , Z_SYNC_FLUSH , Z_FINISH , Z_BLOCK
or
.Dv Z_TREES .
.Dv Z_SYNC_FLUSH
requests that
.Fn inflate
flush as much output as possible to the output buffer.
.Dv Z_BLOCK
requests that
.Fn inflate
stop if and when it gets to the next deflate block boundary.
When decoding the zlib or gzip format, this will cause
.Fn inflate
to return immediately after the header and before the first block.
When doing a raw inflate,
.Fn inflate
will go ahead and process the first block,
and will return when it gets to the end of that block,
or when it runs out of data.
.Pp
The
.Dv Z_BLOCK
option assists in appending to or combining deflate streams.
Also to assist in this, on return
.Fn inflate
will set strm-\*(Gtdata_type to the number of unused bits in the last byte
taken from strm-\*(Gtnext_in, plus 64 if
.Fn inflate
is currently decoding the last block in the deflate stream, plus 128 if
.Fn inflate
returned immediately after decoding an end-of-block code or decoding
the complete header up to just before the first byte of the deflate stream.
The end-of-block will not be indicated until all of the uncompressed data
from that block has been written to strm-\*(Gtnext_out.
The number of unused bits may in general be greater than seven,
except when bit 7 of data_type is set,
in which case the number of unused bits will be less than eight.
.Va data_type
is set as noted here every time
.Fn inflate
returns for all flush options, and so can be used to determine the amount of
currently consumed input in bits.
.Pp
The
.Dv Z_TREES
option behaves as
.Dv Z_BLOCK
does, but it also returns when the end of each deflate block header is reached,
before any actual data in that block is decoded. This allows the caller to
determine the length of the deflate block header for later use in random access
within a deflate block. 256 is added to the value of
.Va strm-\*(Gtdata_type
when
.Fn inflate
returns immediately after reaching the end of the deflate block header.
.Pp
.Fn inflate
should normally be called until it returns
.Dv Z_STREAM_END
or an error.
However if all decompression is to be performed in a single step
.Pq a single call to inflate ,
the parameter
.Fa flush
should be set to
.Dv Z_FINISH .
In this case all pending input is processed and all pending output is flushed;
.Va avail_out
must be large enough to hold all of the uncompressed data for the operation to
complete.
.Po
The size of the uncompressed data may have been saved by the compressor for
this purpose.
.Pc
The use of
.Dv Z_FINISH
is not required to perform an inflation in one step. However it may be used to
inform
.Fn inflate
that a faster approach can be used for the single
.Fn inflate
call.
.Dv Z_FINISH
also informs inflate to not maintain a sliding window if the stream completes,
which reduces inflate's memory footprint. If the stream does not complete,
either because not all of the stream is provided or not enough output space is
provided, then a sliding window will be allocated and
.Fn inflate
can be called again to continue the operation as if
.Dv Z_NO_FLUSH
had been used.
.Pp
In this implementation,
.Fn inflate
always flushes as much output as possible to the output buffer,
and always uses the faster approach on the first call.
So the effects of the
.Fa flush
parameter in this implementation are on the return value of
.Fn inflate
as noted below, when 
.Fn inflate
returns early when
.Dv Z_BLOCK
or
.Dv Z_TREES
is used, and when
.Fn inflate
avoids the allocation of memory for a sliding window when
.Dv Z_FINISH
is used.
.Pp
If a preset dictionary is needed after this call
.Po
see
.Fn inflateSetDictionary
below
.Pc ,
.Fn inflate
sets strm-\*(Gtadler to the Adler-32 checksum of the dictionary
chosen by the compressor and returns Z_NEED_DICT; otherwise it sets
strm-\*(Gtadler to the Adler-32 checksum of all output produced so far
.Po
that is,
.Va total_out
bytes
.Pc
and returns
.Dv Z_OK , Z_STREAM_END
or an error code as described below.
At the end of the stream,
.Fn inflate
checks that its computed Adler-32 checksum is equal to that saved by
the compressor and returns
.Dv Z_STREAM_END
only if the checksum is correct.
.Pp
.Fn inflate
can decompress and check either zlib-wrapped or gzip-wrapped deflate data.
The header type is detected automatically, if requested when initializing with
.Fn inflateInit2 .
Any information contained in the gzip header is not retained,
so applications that need that information should instead use raw inflate; see
.Fn inflateInit2
below, or
.Fn inflateBack
and perform their own processing of the gzip header and trailer. When
processing gzip-wrapped deflate data, strm-\*(Gtadler32 is set to the CRC-32 of
the output producted so far. The CRC-32 is checked against the gzip trailer.
.Pp
.Fn inflate
returns
.Dv Z_OK
if some progress has been made
.Pq more input processed or more output produced ,
.Dv Z_STREAM_END
if the end of the compressed data has been reached and all uncompressed output
has been produced,
.Dv Z_NEED_DICT
if a preset dictionary is needed at this point,
.Dv Z_DATA_ERROR
if the input data was corrupted
.Po
input stream not conforming to the
.Nm zlib
format or incorrect check value
.Pc ,
.Dv Z_STREAM_ERROR
if the stream structure was inconsistent
.Po
for example
.Va next_in
or
.va next_out
was
.Dv Z_NULL
.Pc ,
.Dv Z_MEM_ERROR
if there was not enough memory,
.Dv Z_BUF_ERROR
if no progress is possible or if there was not enough room in the output buffer
when
.Dv Z_FINISH
is used.
Note that
.Dv Z_BUF_ERROR
is not fatal, and
.Fn inflate
can be called again with more input and more output space
to continue compressing.
If
.Dv Z_DATA_ERROR
is returned, the application may then call
.Fn inflateSync
to look for a good compression block if a partial recovery
of the data is desired.
.It Xo
.Ft int
.Fn inflateEnd "z_streamp strm" ;
.Xc
All dynamically allocated data structures for this stream are freed.
This function discards any unprocessed input and does not flush any
pending output.
.Pp
.Fn inflateEnd
returns
.Dv Z_OK
if successful, or
.Dv Z_STREAM_ERROR
if the stream state was inconsistent.
In the error case,
.Va msg
may be set but then points to a static string
.Pq which must not be deallocated .
.El
.Sh ADVANCED FUNCTIONS
The following functions are needed only in some special applications.
.Bl -tag -width Ds
.It Xo
.Ft int
.Fn deflateInit2 "z_streamp strm" "int level" "int method" "int windowBits" "int memLevel" "int strategy" ;
.Xc
.Pp
This is another version of
.Fn deflateInit
with more compression options.
The fields
.Va next_in ,
.Va zalloc ,
.Va zfree ,
and
.Va opaque
must be initialized before by the caller.
.Pp
The
.Fa method
parameter is the compression method.
It must be
.Dv Z_DEFLATED
in this version of the library.
.Pp
The
.Fa windowBits
parameter is the base two logarithm of the window size
.Pq the size of the history buffer .
It should be in the range 8..15 for this version of the library.
Larger values of this parameter result in better compression
at the expense of memory usage.
The default value is 15 if
.Fn deflateInit
is used instead.
.Pp
.Fa windowBits
can also be \-8..\-15 for raw deflate.
In this case, -windowBits determines the window size.
.Fn deflate
will then generate raw deflate data with no zlib header or trailer,
and will not compute an Adler-32 check value.
.Pp
.Fa windowBits
can also be greater than 15 for optional gzip encoding.
Add 16 to
.Fa windowBits
to write a simple gzip header and trailer around the
compressed data instead of a zlib wrapper.
The gzip header will have no file name, no extra data, no comment,
no modification time
.Pq set to zero ,
no header crc, and the operating system will be set to 255
.Pq unknown .
If a gzip stream is being written,
strm-\*(Gtadler is a crc32 instead of an adler32.
.Pp
The
.Fa memLevel
parameter specifies how much memory should be allocated
for the internal compression state.
memLevel=1 uses minimum memory but is slow and reduces compression ratio;
memLevel=9 uses maximum memory for optimal speed.
The default value is 8.
See
.Aq Pa zconf.h
for total memory usage as a function of
.Fa windowBits
and
.Fa memLevel .
.Pp
The
.Fa strategy
parameter is used to tune the compression algorithm.
Use the value
.Dv Z_DEFAULT_STRATEGY
for normal data;
.Dv Z_FILTERED
for data produced by a filter
.Pq or predictor ;
.Dv Z_HUFFMAN_ONLY
to force Huffman encoding only
.Pq no string match ,
or
.Dv Z_RLE
to limit match distances to one
.Pq run-length encoding .
Filtered data consists mostly of small values with a
somewhat random distribution.
In this case, the compression algorithm is tuned to compress them better.
The effect of
.Dv Z_FILTERED
is to force more Huffman coding and less string matching;
it is somewhat intermediate between
.Dv Z_DEFAULT_STRATEGY
and
.Dv Z_HUFFMAN_ONLY .
.Dv Z_RLE
is designed to be almost as fast as
.Dv Z_HUFFMAN_ONLY ,
but gives better compression for PNG image data.
The
.Fa strategy
parameter only affects the compression ratio but not the correctness of the
compressed output, even if it is not set appropriately.
.Dv Z_FIXED
prevents the use of dynamic Huffman codes,
allowing for a simpler decoder for special applications.
.Pp
.Fn deflateInit2
returns
.Dv Z_OK
if successful,
.Dv Z_MEM_ERROR
if there was not enough memory,
.Dv Z_STREAM_ERROR
if any parameter is invalid
.Pq such as an invalid method ,
or
.Dv Z_VERSION_ERROR if the zlib library version
.Pq zlib_version
is incompatible with the version assumed by the caller
.Po
.Dv ZLIB_VERSION
.Pc .
.Va msg
is set to null if there is no error message.
.Fn deflateInit2
does not perform any compression: this will be done by
.Fn deflate .
.It Xo
.Ft int
.Fn deflateSetDictionary "z_streamp strm" "const Bytef *dictionary" "uInt dictLength" ;
.Xc
.Pp
Initializes the compression dictionary from the given byte sequence
without producing any compressed output.
When using the zlib format, this function must be called immediately after
.Fn deflateInit ,
.Fn deflateInit2 ,
or
.Fn deflateReset ,
and before any call to
.Fn deflate .
When doing raw deflate, this function must be called either before any call of
.Fn deflate ,
or immediately after the completion of a deflate block, i.e. after all input has been
consumed and all output has been delivered when using any of the flush
options
.Dv Z_BLOCK ,
.Dv Z_PARTIAL_FLUSH ,
.Dv Z_SYNC_FLUSH ,
or
.Dv Z_FULL_FLUSH .
The compressor and decompressor must use exactly the same dictionary
.Po
see
.Fn inflateSetDictionary
.Pc .
.Pp
The dictionary should consist of strings
.Pq byte sequences
that are likely to be encountered later in the data to be compressed,
with the most commonly used strings preferably put towards
the end of the dictionary.
Using a dictionary is most useful when the data to be compressed is short
and can be predicted with good accuracy;
the data can then be compressed better than with the default empty dictionary.
.Pp
Depending on the size of the compression data structures selected by
.Fn deflateInit
or
.Fn deflateInit2 ,
a part of the dictionary may in effect be discarded,
for example if the dictionary is larger than the window size provided in
.Fn deflateInit
or
.Fn deflateInit2 .
Thus the strings most likely to be useful should be
put at the end of the dictionary, not at the front.
In addition, the current implementation of
.Fn deflate
will use at most the window size minus 262 bytes of the provided dictionary.
.Pp
Upon return of this function, strm-\*(Gtadler is set to the Adler-32 value
of the dictionary; the decompressor may later use this value to determine
which dictionary has been used by the compressor.
.Po
The Adler-32 value applies to the whole dictionary even if only a subset of the
dictionary is actually used by the compressor.
.Pc
If a raw deflate was requested, then the Adler-32 value is not computed and
strm-\*(Gtadler is not set.
.Pp
.Fn deflateSetDictionary
returns
.Dv Z_OK
if successful,
or
.Dv Z_STREAM_ERROR
if a parameter is invalid
.Po
e.g. dictionary being
.Dv Z_NULL
.Pc
or the stream state is inconsistent
.Po
for example if
.Fn deflate
has already been called for this stream or not at a block boundary for raw
deflate
.Pc .
.Fn deflateSetDictionary
does not perform any compression: this will be done by
.Fn deflate .
.It Xo
.Ft int
.Fn deflateCopy "z_streamp dest" "z_streamp source" ;
.Xc
.Pp
The
.Fn deflateCopy
function sets the destination stream as a complete copy of the source stream.
.Pp
This function can be useful when several compression strategies will be
tried, for example when there are several ways of pre-processing the input
data with a filter.

lib/libz/compress.c
/*	$OpenBSD: compress.c,v 1.6 2005/07/20 15:56:41 millert Exp $	*/
/* compress.c -- compress a memory buffer
 * Copyright (C) 1995-2005 Jean-loup Gailly.
 * For conditions of distribution and use, see copyright notice in zlib.h
 */


#define ZLIB_INTERNAL
#include "zlib.h"

/* ===========================================================================
     Compresses the source buffer into the destination buffer. The level
   parameter has the same meaning as in deflateInit.  sourceLen is the byte
   length of the source buffer. Upon entry, destLen is the total size of the
   destination buffer, which must be at least 0.1% larger than sourceLen plus
   12 bytes. Upon exit, destLen is the actual size of the compressed buffer.

     compress2 returns Z_OK if success, Z_MEM_ERROR if there was not enough
   memory, Z_BUF_ERROR if there was not enough room in the output buffer,
   Z_STREAM_ERROR if the level parameter is invalid.
*/
int ZEXPORT compress2 (dest, destLen, source, sourceLen, level)
    Bytef *dest;
    uLongf *destLen;
    const Bytef *source;
    uLong sourceLen;
    int level;
{
    z_stream stream;
    int err;

    stream.next_in = (z_const Bytef *)source;
    stream.avail_in = (uInt)sourceLen;
#ifdef MAXSEG_64K
    /* Check for source > 64K on 16-bit machine: */
    if ((uLong)stream.avail_in != sourceLen) return Z_BUF_ERROR;
#endif
    stream.next_out = dest;
    stream.avail_out = (uInt)*destLen;
    if ((uLong)stream.avail_out != *destLen) return Z_BUF_ERROR;

    stream.zalloc = (alloc_func)0;
    stream.zfree = (free_func)0;
    stream.opaque = (voidpf)0;

    err = deflateInit(&stream, level);
    if (err != Z_OK) return err;

    err = deflate(&stream, Z_FINISH);
    if (err != Z_STREAM_END) {
        deflateEnd(&stream);
        return err == Z_OK ? Z_BUF_ERROR : err;
    }
    *destLen = stream.total_out;

    err = deflateEnd(&stream);
    return err;
}

/* ===========================================================================
 */
int ZEXPORT compress (dest, destLen, source, sourceLen)
    Bytef *dest;
    uLongf *destLen;
    const Bytef *source;
    uLong sourceLen;
{
    return compress2(dest, destLen, source, sourceLen, Z_DEFAULT_COMPRESSION);
}

/* ===========================================================================
     If the default memLevel or windowBits for deflateInit() is changed, then
   this function needs to be updated.
 */
uLong ZEXPORT compressBound (sourceLen)
    uLong sourceLen;
{
    return sourceLen + (sourceLen >> 12) + (sourceLen >> 14) +
           (sourceLen >> 25) + 13;
}
